tests/Exercise/BirdCountTest.elm

```diff
@@ -0,0 +1,170 @@
+module Exercise.BirdCountTest exposing (tests)
+
+import Comment exposing (Comment, CommentType(..))
+import Dict
+import Exercise.BirdCount as BirdCount
+import Review.Rule exposing (Rule)
+import Review.Test
+import RuleConfig
+import Test exposing (Test, describe, test)
+import TestHelper
+
+
+tests : Test
+tests =
+    describe "BirdCountTest"
+        [ exemplar
+        , usingList
+        ]
+
+
+rules : List Rule
+rules =
+    BirdCount.ruleConfig |> .rules |> List.map RuleConfig.analyzerRuleToRule
+
+
+exemplar : Test
+exemplar =
+    test "should not report anything for the example solution" <|
+        \() ->
+            TestHelper.expectNoErrorsForRules rules
+                """
+module BirdCount exposing (busyDays, hasDayWithoutBirds, incrementDayCount, today, total)
+
+
+today : List Int -> Maybe Int
+today counts =
+    case counts of
+        [] ->
+            Nothing
+
+        head :: _ ->
+            Just head
+
+
+incrementDayCount : List Int -> List Int
+incrementDayCount counts =
+    case counts of
+        [] ->
+            [ 1 ]
+
+        head :: tail ->
+            (head + 1) :: tail
+
+
+hasDayWithoutBirds : List Int -> Bool
+hasDayWithoutBirds counts =
+    case counts of
+        [] ->
+            False
+
+        0 :: _ ->
+            True
+
+        _ :: tail ->
+            hasDayWithoutBirds tail
+
+
+total : List Int -> Int
+total counts =
+    case counts of
+        [] ->
+            0
+
+        head :: tail ->
+            head + total tail
+
+
+busyDays : List Int -> Int
+busyDays counts =
+    case counts of
+        [] ->
+            0
+
+        head :: tail ->
+            if head >= 5 then
+                1 + busyDays tail
+
+            else
+                busyDays tail
+"""
+
+
+usingList : Test
+usingList =
+    let
+        comment =
+            Comment "elm.bird-count.do_not_use_list" Essential Dict.empty
+    in
+    describe "solutions that use the List function" <|
```

Comment:
I don't think I would test all the different uses of `List`, being as the rule disallows the module completely, and `AnyFromExternalModule` and similar are already tested, but its not a hill I am going to die on.

Reply:
The point of these tests is not to check that the rule works properly, like you said this has been tested a lot by now, the point is to show examples of "wrong" solutions that justify why we have those rules in place.

With these code examples, you can conclude "right, if someone writes a solution like that, they won't learn anything about recursion." Conversely, if you are not able to come up with an reasonable example of a solution breaking a rule, then maybe the rule isn't meaningful.

Reply:
We can agree to differ on this, but I do take your point.